src/drawer/drawer.service.ts

```diff
@@ -1,65 +1,116 @@
+export class DrawerService<T = unknown, C = unknown, R = any> {
   private overlayRef: OverlayRef;
-  private readonly unsubscribe$ = new Subject<void>();
+  options: DrawerOptions<T, C>;
+  drawerRef: DrawerRef<R>;
+  invisible$ = new Subject();
+  private drawerCpt: ComponentRef<DrawerInternalComponent<T, C>>;
 
   constructor(private readonly overlay: Overlay) {}
 
-  open<T = unknown>(options: DrawerOptions<T>) {
-    this.drawerRef?.instance?.dispose();
-    this.createDrawer();
+  open(options: DrawerOptions<T, C>) {
     this.updateOptions(options);
-    return this.drawerRef?.instance;
+    this.createOverlay();
+    this.createDrawer();
+    this.drawerRef = new DrawerRef(this.drawerCpt.instance);
+    this.drawerRef.open();
+
+    return this.drawerRef;
   }
 
-  updateOptions(options: DrawerOptions): void {
-    Object.assign(this.drawerRef.instance, options);
+  updateOptions(options: DrawerOptions<T, C>): void {
+    this.options = options;
   }
 
-  private createDrawer(): void {
-    this.overlayRef = this.overlay.create();
-    this.drawerRef = this.overlayRef.attach(
-      new ComponentPortal(DrawerComponent),
-    );
-    this.drawerRef.instance.drawerViewInit
-      .pipe(takeUntil(this.unsubscribe$))
-      .subscribe(() => {
-        this.drawerRef.instance.open();
-      });
+  private createOverlay() {
+    if (!this.overlayRef) {
+      this.overlayRef = this.overlay.create(this.getOverlayConfig());
+    }
 
-    this.drawerRef.instance.afterClosed
-      .pipe(takeUntil(this.unsubscribe$))
-      .subscribe(() => {
-        this.overlayRef.dispose();
-        this.drawerRef = null;
-        this.unsubscribe$.next();
-        this.unsubscribe$.complete();
+    this.overlayRef
+      .outsidePointerEvents()
+      .pipe(takeUntil(this.invisible$))
+      .subscribe(event => {
+        // 判断鼠标点击事件的 target 是否为 overlay-container 的子节点，如果是，则不关闭 drawer。
+        // 为了避免点击 drawer 里的 tooltip 后 drawer 被关闭。
+        if (
+          this.overlayRef &&
+          this.options.hideOnClickOutside &&
+          event.target instanceof Node &&
+          !this.overlayRef.hostElement?.parentNode?.contains(event.target)
+        ) {
+          event.stopPropagation();
+          event.preventDefault();
+          this.drawerRef.close();
+        }
       });
+
+    this.overlayRef.getConfig().scrollStrategy.enable();
+    if (this.options.mask) {
+      // Issues: https://github.com/angular/components/issues/10841
+      // scrollStrategy 为 Block 时，若创建 Overlay 时，高度不足以出现滚动，则 scrollStrategy 不会生效
+      fromEvent(window, 'resize')
+        .pipe(
+          debounceTime(100),
+          filter(
+            () => document.documentElement.scrollHeight > window.innerHeight,
+          ),
+          takeUntil(this.invisible$),
+        )
+        .subscribe(() => {
+          this.overlayRef.getConfig().scrollStrategy.enable();
+        });
+    }
+  }
+
+  private createDrawer() {
+    if (this.drawerCpt) {
```

Comment:
```suggestion
    if (this.drawerInstance) {
```

`Cpt` 是 `Component` 缩写吗？除了一些共识度非常高的尽量少用缩写吧